gcp_authn: Add the doc and release note for gcp_authn_filter

tyxia opened the pull request:
Signed-off-by: Tianyu Xia <tyxia@google.com>

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

phlax reviewed (changes requested):
first pass review inline...

phlax commented:
/docs

tyxia commented:
/unassign @markdroth 

No API change, added the API reviewer by mistake (rebase command)

tyxia commented:
> if you rebase to `upstream/main` and force push it should resolve the merge mash

Thank you for early review(I meant to sent out as WIP draft) and hints! For the merge mess, I was trying Github UI suggested edit feature which caused the DCO sign-off error and then I followed the suggested instruction which then caused the merge mess here :(.
It should be resolved now other than expanded `participants` (i.e., expanded noise...), which I am not sure how to/if possible to remove.

Also, I am addressing your review comments and updating doc with latest feature. I will switch it to `ready for review` when that is done.

phlax commented:
> Thank you for early review

np, just letting you know that its ok to force-push on a WIP PR - all reviews have been on the basis of it being WIP 

re merge mash - its a pita 8/ and it flags loads of things unneessarily - so was just clarifying how to resolve

phlax commented on docs/root/configuration/http/http_filters/_include/gcp-authn-filter-configuration.yaml:
```diff
@@ -0,0 +1,80 @@
+              timeout:
+                seconds: 10
+          http_filters:
+          - name: "envoy.filters.http.router"
```
this needs a `typed_config` 

there are 2 config validation checks - the first is more like static linting - this one, that has failed, starts a server with the binary+config. It seems the first check is not enforcing everything it should

tyxia replied:
Thanks for the review! PTAL.

tyxia commented:
/retest

adisuissa reviewed (commented):
Thanks for working on this.
Does this mean this work is now complete? If so, please add a release note.

phlax commented:
@tyxia could you merge main and ill review again

/wait

phlax commented:
@tyxia if you can merge main i would be happy to review again

phlax commented on docs/root/configuration/http/http_filters/_include/gcp-authn-filter-configuration.yaml:
```diff
@@ -0,0 +1,82 @@
+      typed_filter_metadata:
+        envoy.filters.http.gcp_authn:
+          "@type": type.googleapis.com/envoy.extensions.filters.http.gcp_authn.v3.Audience
+          url: http:#test.com
```
this seems strange - can we use a real/fake url instead ?

phlax commented on docs/root/configuration/http/http_filters/_include/gcp-authn-filter-configuration.yaml:
```diff
@@ -0,0 +1,82 @@
+          - name: envoy.filters.http.router
+            typed_config:
+              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+          access_log:
```
i think this is not necessary as this is not what is being shown

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+The filter configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig>` has three fields:
+
+* Field ``http_uri`` specifies the HTTP URI for fetching the from `Google Compute Engine(GCE) Metadata Server <https://cloud.google.com/compute/docs/metadata/overview>`_. The URL format is ``http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity?audience=[AUDIENCE]``. The ``AUDIENCE`` field is provided by configuration, please see more details below.
```
```suggestion
* ``http_uri`` specifies the HTTP URI for fetching the from `Google Compute Engine(GCE) Metadata Server <https://cloud.google.com/compute/docs/metadata/overview>`_. The URL format is ``http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity?audience=[AUDIENCE]``. The ``AUDIENCE`` field is provided by configuration, please see more details below.
```

i would remove these "Field " prefixes as i think it would be simpler (and it says above "...three fields:")

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+* Field ``http_uri`` specifies the HTTP URI for fetching the from `Google Compute Engine(GCE) Metadata Server <https://cloud.google.com/compute/docs/metadata/overview>`_. The URL format is ``http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity?audience=[AUDIENCE]``. The ``AUDIENCE`` field is provided by configuration, please see more details below.
+
+* Field ``retry_policy`` specifies the retry policy if fetching tokens failed. This field is optional. If it is not configured, the filter will be fail-closed (i.e., reject the requests).
```
```suggestion
* Field ``retry_policy`` specifies the retry policy if fetching tokens failed. This field is optional. If it is not configured, the filter will be fail-closed (i.e. reject the requests).
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+* Field ``cache_config`` specifies the configuration for the token cache which is used to avoid the duplicated queries to GCE metadata server for the same request.
+
+The audience configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.Audience>` is the URL of the destionation service, which is the receving service that calling service is invoking. This information is provided through cluster metadata :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`
```
```suggestion
The audience configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.Audience>` is the URL of the destination service, which is the receiving service that the calling service is invoking. This information is provided through cluster metadata :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+The audience configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.Audience>` is the URL of the destionation service, which is the receving service that calling service is invoking. This information is provided through cluster metadata :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`
+
+The token cache configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.TokenCacheConfig>` is used to avoid the redundant queries to authentication server (GCE metadata server in the context of this filter) for duplicated tokens.
```
```suggestion
The token cache configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.TokenCacheConfig>` is used to avoid the redundant queries to the authentication server (GCE metadata server in the context of this filter) for duplicated tokens.
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+Configuration example
+---------------------
+Resouce configuration example:
```
```suggestion
Resource configuration example:
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+.. literalinclude:: _include/gcp-authn-filter-configuration.yaml
+   :language: yaml
+   :lines: 39-80
```
i think this needs adusting

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+The token cache configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.TokenCacheConfig>` is used to avoid the redundant queries to authentication server (GCE metadata server in the context of this filter) for duplicated tokens.
+
+Configuration example
+---------------------
```
add an empty line beneath here please

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+
+* Field ``retry_policy`` specifies the retry policy if fetching tokens failed. This field is optional. If it is not configured, the filter will be fail-closed (i.e., reject the requests).
+
+* Field ``cache_config`` specifies the configuration for the token cache which is used to avoid the duplicated queries to GCE metadata server for the same request.
```
```suggestion
* Field ``cache_config`` specifies the configuration for the token cache which is used to avoid the duplicated queries to the GCE metadata server for the same request.
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,42 @@
+   :lineno-start: 39
+   :caption: :download:`gcp-authn-filter-configuration.yaml <_include/gcp-authn-filter-configuration.yaml>`
+
+Http filter configuration example:
```
```suggestion
HTTP filter configuration example:
```

phlax reviewed (changes requested):
thanks for working on this @tyxia 

overall i would try to simplify the config any way we can - ie remove any defaults if they are not specifically documented in the text

im wondering about using ip6 here - all of the other examples use ip4 so i would suggest using that if possible (maybe we should default to ip6 but i think it needs to be consistent)

we dont enforce line length for rst but having very long lines makes it harder to review and change so i would split some of the lines (if there is no blank line in between they will be run together anyway)

tyxia commented:
> thanks for working on this @tyxia
> 
> overall i would try to simplify the config any way we can - ie remove any defaults if they are not specifically documented in the text
> 
> im wondering about using ip6 here - all of the other examples use ip4 so i would suggest using that if possible (maybe we should default to ip6 but i think it needs to be consistent)
> 
> we dont enforce line length for rst but having very long lines makes it harder to review and change so i would split some of the lines (if there is no blank line in between they will be run together anyway)

Thank you very much for the thorough review, @phlax. Those configurations are just dumped from my integration tests and I didn't modify any fields by hand. I will try to simplify them. Also, I will address the other review comments later.

phlax commented:
i think not stale

tyxia reviewed (commented):
Thanks for the review @phlax 

Those configuration examples are just dumped from my integration tests so they have some extra fields. I have simplified them and please take another look.

tyxia commented:
> Thanks for working on this. Does this mean this work is now complete? If so, please add a release note.

Thanks for the review, Adi! Yes, it has been used by several customers already. I have added the release note. 
PTAL, thanks!

adisuissa commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,46 @@
+The audience configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.Audience>` is the URL of the destination service,
+which is the receiving service that the calling service is invoking. This information is provided through cluster metadata :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`
+
+The token cache configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.TokenCacheConfig>` is used to avoid the redundant queries to
```
```suggestion
The token cache configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.TokenCacheConfig>` is used to avoid redundant queries to
```

tyxia commented:
> Thanks! Looks good overall, left a couple of minor comments.
Thanks! PTAL

tyxia commented:
> Thanks! Looks good overall, left a couple of minor comments.

Thanks for the review! PTAL

tyxia commented:
/retest

tyxia commented:
/retest

adisuissa reviewed (dismissed):
LGTM, thanks!

adisuissa commented:
@phlax for a final pass and merge

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,46 @@
-This filter is used to fetch authentication tokens from GCP compute metadata server(https://cloud.google.com/run/docs/securing/service-identity#identity_tokens).
-In multiple services architecture where these services likely need to communicate withe each other, authenticating service-to-service(https://cloud.google.com/run/docs/authenticating/service-to-service) is required because many of these services may be private and require credentials for access.
+This filter is used to fetch authentication tokens from `Google Compute Engine(GCE) metadata server <https://cloud.google.com/compute/docs/metadata/overview>`_.
+In multiple services architecture where these services likely need to communicate with each other,
```
```suggestion
In a multiple services architecture where the services need to communicate with each other,
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,46 @@
+
+* ``retry_policy`` specifies the retry policy if fetching tokens failed. This field is optional. If it is not configured, the filter will be fail-closed (i.e., reject the requests).
+
+* ``cache_config`` specifies the configuration for the token cache which is used to avoid the duplicated queries to GCE metadata server for the same request.
```
```suggestion
* ``cache_config`` specifies the configuration for the token cache which is used to avoid duplicated queries to GCE metadata server for the same request.
```

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,46 @@
+* ``cache_config`` specifies the configuration for the token cache which is used to avoid the duplicated queries to GCE metadata server for the same request.
+
+The audience configuration :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.gcp_authn.v3.Audience>` is the URL of the destination service,
+which is the receiving service that the calling service is invoking. This information is provided through cluster metadata :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`
```
```suggestion
which is the receiving service that the calling service is invoking. This information is provided through cluster :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`
```

tyxia replied:
I guess `cluster metadata ` was a bit confusing. I changed to `cluster's metadata field`.

phlax commented on docs/root/configuration/http/http_filters/gcp_authn_filter.rst:
```diff
@@ -2,10 +2,46 @@
-In multiple services architecture where these services likely need to communicate withe each other, authenticating service-to-service(https://cloud.google.com/run/docs/authenticating/service-to-service) is required because many of these services may be private and require credentials for access.
+This filter is used to fetch authentication tokens from `Google Compute Engine(GCE) metadata server <https://cloud.google.com/compute/docs/metadata/overview>`_.
+In multiple services architecture where these services likely need to communicate with each other,
+`authenticating service-to-service <https://cloud.google.com/run/docs/authenticating/service-to-service>`_ is needed because many of these services may be private and require credentials for access.
```
```suggestion
`authenticating service-to-service <https://cloud.google.com/run/docs/authenticating/service-to-service>`_ is needed where services are private and require credentials for access.
```

phlax reviewed (commented):
Thanks @tyxia - couple of nits and then should be gtm

tyxia commented:
> Thanks @tyxia - couple of nits and then should be gtm

Thank you for review! I have addressed your comment. PTAL.

The CI failure should not to be related to PR itself. Hopefully the retest could resolve it.

tyxia commented:
/retest

phlax reviewed (approved):
lgtm, thanks @tyxia
